src/core/request_handler.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2020, Arm Limited, All Rights Reserved
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//          http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![allow(dead_code)]
+
+use parsec_interface::requests::{Request, Response, Result};
+use std::os::unix::net::UnixStream;
+use std::path::PathBuf;
+use std::time::Duration;
+
+const DEFAULT_MAX_BODY_SIZE: usize = 1 << 31;
+const DEFAULT_SOCKET_PATH: &str = "/tmp/security-daemon-socket";
+
+/// Low level client structure to send a `Request` and get a `Response`.
+#[derive(Clone, Debug)]
+pub struct RequestHandler {
+    pub max_body_size: usize,
+    pub timeout: Option<Duration>,
+    pub socket_path: PathBuf,
+}
+
+impl RequestHandler {
+    /// Send a request and get a response.
+    pub fn send_request(&self, request: Request) -> Result<Response> {
+        // Try to connect once, wait for a timeout until trying again.
+        let mut stream =
+            UnixStream::connect(&self.socket_path).expect("Failed to connect to Unix socket");
+
+        stream
+            .set_read_timeout(self.timeout)
+            .expect("Failed to set read timeout for stream");
+        stream
+            .set_write_timeout(self.timeout)
+            .expect("Failed to set write timeout for stream");
```

Comment:
Maybe the socket connection + the timeout setting can be done only once when building the `RequestHandler`? So that after it is faster so send requests.

Reply:
Not really, this is a problem we encountered last time - one stream = one E2E connection. Once the response is sent, the service closes its side and writing to the stream results in a `broken pipe` error.

Reply:
(Also not closing the connection on the service side would probably be a memory leak)

Reply:
Oh yes sorry you are right! I forgot about that one 😃